ENH: Implement Kleene logic for BooleanArray

TomAugspurger opened the pull request:
xref #29556 

I have a few TODOs, and a few tests that I need to unxfail. Putting this up now so that @jorisvandenbossche can take a look.

jorisvandenbossche committed:
ENH: add BooleanArray extension array (#29555)

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -576,17 +581,38 @@ def logical_method(self, other):
+                    mask[~other & ~omask] = False
+                elif lib.is_scalar(other) and np.isnan(other):  # TODO(NA): change to NA
+                    mask[:] = True
+                # Do we ever assume that masked values are False?
```
@jorisvandenbossche do we assume / promise that `self._data[self._mask]` is all False?

jorisvandenbossche replied:
No, there is currently no guarantee on the value in `_data` that is masked

TomAugspurger commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -391,13 +391,101 @@ def test_scalar(self, data, all_logical_operators):
     def test_array(self, data, all_logical_operators):
         op_name = all_logical_operators
+        if "or" in op_name:
+            pytest.skip("confusing")
```
This will obviously be addressed. Just need to understand the test.

TomAugspurger commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -391,13 +391,101 @@ def test_scalar(self, data, all_logical_operators):
         other = pd.Series([True] * len(data), dtype="boolean")
         self._compare_other(data, op_name, other)
 
+    def test_kleene_or(self):
```
A careful review of these new test cases would be greatly appreciated. I've tried to make them as clear as possible, while covering all the cases.

jorisvandenbossche replied:
I went through the tests, very clear, added a few comments, for the rest looks good to me!

TomAugspurger committed:
move

TomAugspurger committed:
doc fixup

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into boolean-array-kleene

TomAugspurger committed:
working

jorisvandenbossche commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -391,13 +391,101 @@ def test_scalar(self, data, all_logical_operators):
+
+    def test_kleene_or_scalar(self):
+        a = pd.array([True, False, None], dtype="boolean")
+        result = a | np.nan  # TODO: pd.NA
```
But that gives the question: what do we want `a | np.nan` to do? 
Standard python and numpy both raise a TypeError for logical operations with floats

jorisvandenbossche commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -391,13 +391,101 @@ def test_scalar(self, data, all_logical_operators):
+
+        result = np.nan | a  # TODO: pd.NA
+        tm.assert_extension_array_equal(result, expected)
+
```
Can you also test the case where the scalar is True or False here? (that is still related to Kleene, since there is a missing value in the array)

jorisvandenbossche replied:
Or integrate with test case below?

jorisvandenbossche commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -391,13 +391,101 @@ def test_scalar(self, data, all_logical_operators):
+    def test_kleene_and_scalar(self):
+        a = pd.array([True, False, None], dtype="boolean")
+        result = a & np.nan  # TODO: pd.NA
+        expected = pd.array([None, None, None], dtype="boolean")
```
this needs to be [None, False, None] (or once replaced with pd.NA ?)

jorisvandenbossche reviewed (commented):
I started adding some docs in https://github.com/pandas-dev/pandas/pull/29597 as well (on the NA scalar), so need to think about potential overlap. Now, I think we need to explain it in both places anyhow.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into boolean-array-kleene

TomAugspurger committed:
updates

TomAugspurger committed:
updates

TomAugspurger commented:
Sorry @jorisvandenbossche, I pushed a somewhat major refactor after you reviewed :/ Should be done now.

High-level overview:

* Split into dedicated methods `kleene_or`, `kleene_and`, and `kleene_xor`, rather than doing everything in `_create_logical_method`.
* Removed the `_compare_other`, `test_scalar`, and `test_array` tests. Since there were parameterized over all the ops, we would have needed to essentially re-implement the kleene logic in `_compare_other` to get the masking right.
* Added tests for each op against both arrays and scalars (True, False, np.nan)

My main concern right now is that I may be assuming that masked values are false in a few places.

---

One important comment was buried

> But that gives the question: what do we want a | np.nan to do?

Right now, I've adopted `pd.NA` semantics. I think we do that or raise, I don't have a preference. Easy to support both.

jorisvandenbossche commented:
> My main concern right now is that I may be assuming that masked values are false in a few places.

How would you be assuming that? Is there a place that you "uncover" masked values?

jorisvandenbossche commented:
> Right now, I've adopted pd.NA semantics. I think we do that or raise, I don't have a preference. Easy to support both.

I would maybe rather raise an error then. As otherwise you have a `np.nan` that behaves differently depending on the context.

TomAugspurger commented:
> How would you be assuming that? Is there a place that you "uncover" masked values?

Still thinking through it, but we do things rougly like

```python
result = left & right
...
mask[result] = False
```

i.e. we update the mask based on the result. I'll add some tests where I manually modify the `_data` of masked values to be True.

jorisvandenbossche commented:
> i.e. we update the mask based on the result. 

Ah, I see. Yes, that shouldn't be done. Or it should combine with mask from the original values first, I think?

jorisvandenbossche commented on pandas/core/arrays/boolean.py:
```diff
@@ -740,6 +742,162 @@ def boolean_arithmetic_method(self, other):
+
+    # XXX: verify that this doesn't assume masked values are False!
+    result = left | right
+    mask[result] = False
```
In the end, we want to know where either left or right was a True, right? 
But, that also depends on the mask, as the masked values should not be taken into account. 

So should this be something like `result = (left & ~left_mask) | (right & ~right_mask)`

(although there might be a more performant way with less steps)

jorisvandenbossche replied:
And then the final mask would be something like `(mask_left | mask_right) & ~result`

jorisvandenbossche replied:
But that gives many logical ops to perform one logical op ...

TomAugspurger replied:
Indeed, that's the downside.

For now, I think we just accept that cost. We attempt to ensure that masked values are always False, but that seems difficult to get correct, and becomes obsolete if / when we back these with Pyarrow arrays.

jorisvandenbossche replied:
> We attempt to ensure that masked values are always False, but that seems difficult to get correct

I don't think we should even try to attempt that

TomAugspurger committed:
Raise for NaN

TomAugspurger committed:
added tests for empty

TomAugspurger committed:
added tests for inplace mutation

TomAugspurger committed:
Do not assume masked values are False

TomAugspurger commented:
Another batch of changes

* raise for NaN: https://github.com/pandas-dev/pandas/pull/29842/commits/7f78a64faee7d9be2c88990f3f930fa89fae2d11
* More tests (empty arrays, inplace mutation) https://github.com/pandas-dev/pandas/pull/29842/commits/747e046ed54866498260f5922fdbc44eb0e4a85f, https://github.com/pandas-dev/pandas/pull/29842/commits/36b171b2ff1f01563367c13313572d257d7d67fc
* Ensure we aren't relying on the assumption that masked values are False: https://github.com/pandas-dev/pandas/pull/29842/commits/d0a8cca3d8499e62c5488612a527756e7ad9057e

The tests in https://github.com/pandas-dev/pandas/pull/29842/commits/d0a8cca3d8499e62c5488612a527756e7ad9057e are a bit tricky... Hopefully they're comprehensive and make sense.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into boolean-array-kleene

jreback commented on doc/source/user_guide/boolean.rst:
```diff
@@ -0,0 +1,77 @@
+Kleene Logic
+------------
+
+:class:`arrays.BooleanArray` implements Kleene logic (sometimes called three-value logic) for
```
can you add the wiki reference here

TomAugspurger committed:
mypy

jreback commented on pandas/core/arrays/boolean.py:
```diff
@@ -740,6 +742,171 @@ def boolean_arithmetic_method(self, other):
         return set_function_name(boolean_arithmetic_method, name, cls)
 
 
+def kleene_or(
```
would these be better in a separate module (in case we decide to re-use them) and this becomes less cluttered

TomAugspurger committed:
doc fixups

jorisvandenbossche commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -360,43 +360,200 @@ def get_op_from_name(self, op_name):
+            other = pd.array(other, dtype="boolean")
+
+        # mutate the data inplace
+        a._data[a._mask] = True
```
Instead of mutating here, you can also just create the BooleanArray(values, mask) with the exact values you want

jorisvandenbossche commented on pandas/core/arrays/boolean.py:
```diff
@@ -740,6 +742,172 @@ def boolean_arithmetic_method(self, other):
+        result = left & right
+        # unmask where either left or right is False
+        mask[~left & ~left_mask] = False
+        mask[~right & ~right_mask] = False
```
Similar comment here, I think that something like this can be faster:

```
left_false = ~left & ~left_mask
right_false= ~right & ~right_mask

mask = (left_mask & ~right_false) | (right_mask & ~left_false)
```

(avoiding setitem)

And need to think if we can avoid some `~`

jorisvandenbossche replied:
Further optimization:

```
left_false = ~(left | left_mask)
right_false= ~(right | right_mask)
mask = (left_mask & ~right_false) | (right_mask & ~left_false)
```

Timing comparison:

```
left = np.random.randint(0, 2, 1000).astype(bool)
right = np.random.randint(0, 2, 1000).astype(bool) 
left_mask = np.random.randint(0, 2, 1000).astype(bool) 
right_mask = np.random.randint(0, 2, 1000).astype(bool)
```

```
In [47]: %%timeit 
    ...: mask = left_mask | right_mask 
    ...: mask[~left & ~left_mask] = False 
    ...: mask[~right & ~right_mask] = False 

7.2 µs ± 106 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [58]: %%timeit 
    ...: left_false = ~(left | left_mask) 
    ...: right_false= ~(right | right_mask) 
    ...:  
    ...: mask = (left_mask & ~right_false) | (right_mask & ~left_false)
3.73 µs ± 275 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

jorisvandenbossche replied:
And on bigger arrays, the difference is much bigger, it seems. For 100_000 elements, I get 775 µs vs 45 µs

TomAugspurger replied:
Thanks. I'll also add an asv for these ops.

jreback reviewed (changes requested):
some initial comments, haven't looked at the tests yet

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -740,6 +742,171 @@ def boolean_arithmetic_method(self, other):
+    mask = left_mask
+
+    if right_mask is not None:
+        mask = mask | right_mask
```
I believe so, though I may be wrong...

TomAugspurger replied:
Ah sorry, not needed after using your code. Thanks.

TomAugspurger committed:
Added benchmarks

TomAugspurger committed:
update tests

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into boolean-array-kleene

jorisvandenbossche commented on pandas/core/arrays/boolean.py:
```diff
@@ -740,6 +741,160 @@ def boolean_arithmetic_method(self, other):
+        right_false = ~(right | right_mask)
+        mask = (left_mask & ~right_false) | (right_mask & ~left_false)
+
+    result[mask] = False
```
I don't think we need to do this (we can't rely on that in general anyway, and then this just gives a performance degradation)

jorisvandenbossche commented:
I think a similar optimization as I did for "and" should also be possible for "or"

TomAugspurger committed:
remove unneded setitem

TomAugspurger committed:
optimize

TomAugspurger commented:
> I think a similar optimization as I did for "and" should also be possible for "or"

Done in https://github.com/pandas-dev/pandas/pull/29842/commits/7b9002cf52e93ba49bb425da6abf09b3fb0cf529.

TomAugspurger committed:
comments

TomAugspurger committed:
just do the xor

jorisvandenbossche commented on pandas/core/arrays/boolean.py:
```diff
@@ -773,19 +773,20 @@ def kleene_or(
-        mask = mask | right_mask
+        # output is unknown where (False & NA), (NA & False), (NA & NA)
+        mask = (
+            ((~left & ~left_mask) & right_mask)
```
the "left False" can also be done as `~(left | left_mask)`, which is a little bit faster (one operation less) 

(and same for "right false" below)

jorisvandenbossche commented on doc/source/user_guide/boolean.rst:
```diff
@@ -0,0 +1,80 @@
+
+.. versionadded:: 1.0.0
+
+.. _boolean.klean:
```
```suggestion
.. _boolean.kleene:
```

jbrockmendel commented:
xref #2434?

h-vetinari commented:
Also xref #6528 #22724. Big +1

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into boolean-array-kleene

TomAugspurger committed:
fixup docstring

TomAugspurger committed:
fix label

TomAugspurger committed:
PERF: faster or

jorisvandenbossche commented:
@TomAugspurger it might be best to first finish NA in BooleanArray in https://github.com/pandas-dev/pandas/pull/29961, and then update this PR to also handle the scalar cases?

TomAugspurger commented:
@jorisvandenbossche yes, I think that order makes sense.

jbrockmendel commented:
This may have come up, but FYI, _libs.ops has a few functions that implement kleen-adjacent logic, might be worth using/adapting

TomAugspurger commented:
Are you referring to `vec_binop` and `scalar_binop`? If so, I don't think there's anything there that can be reused. From what I can tell, those just pass through the one value when the other is `NA`.

```python
        if x is None or is_nan(x):
            result[i] = x
        else:
            result[i] = op(x, val)
```

For kleene-logic, whether the NA is propagated depends on the operation being performed, and potentially the non-NA value itself.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into boolean-array-kleene

TomAugspurger committed:
handle pd.NA

TomAugspurger committed:
validate

TomAugspurger committed:
please mypy

TomAugspurger committed:
move to nanops

TomAugspurger commented:
I've moved the `kleene*` functions to nanops. I don't have a strong preference for where they live.

jorisvandenbossche commented:
> I've moved the kleene* functions to nanops. I don't have a strong preference for where they live.

Currently the nanops only hold reductions (there are actually nancmp, but they are not used AFAIK), so not fully sure it necessarily fits there better (certainly if it doesn't use any of the machinery of that file). And it also already a quite long file. 
(maybe we should start a "maskops" file ..)

jorisvandenbossche commented:
But, although my comment above, I also don't care much where it is put ;)

jbrockmendel commented:
nanops is nearly all unary ops, where core.ops is mostly binary ops.  So i think these probably belong in core.ops

jbrockmendel commented on pandas/core/nanops.py:
```diff
@@ -1417,3 +1417,170 @@ def nanpercentile(values, q, axis, na_value, mask, ndim, interpolation):
+
+
+def raise_for_nan(value, method):
+    if lib.is_scalar(value) and isinstance(value, float) and np.isnan(value):
```
the lib.is_scalar is redundant with the isinstance check; you probably just want `lib.is_float`

jbrockmendel commented on pandas/core/nanops.py:
```diff
@@ -1417,3 +1417,170 @@ def nanpercentile(values, q, axis, na_value, mask, ndim, interpolation):
+
+    raise_for_nan(right, method="or")
+
+    if right is libmissing.NA:
```
should libmissing.NA be in the signature for left or right?

TomAugspurger replied:
Added libmissing.NAType.

jbrockmendel commented on pandas/core/nanops.py:
```diff
@@ -1417,3 +1417,170 @@ def nanpercentile(values, q, axis, na_value, mask, ndim, interpolation):
+    left, right : ndarray, NA, or bool
+        The values of the array.
+    left_mask, right_mask : ndarray, optional
+        The masks. When
```
You stopped writing in the middle of a

jbrockmendel commented:
can you add tests checking that the arrays/masks arent getting modified inplace (which is a problem in IntegerArray)

jorisvandenbossche commented:
> can you add tests checking that the arrays/masks arent getting modified inplace (which is a problem in IntegerArray)

@jbrockmendel there are several cases of `# ensure we haven't mutated anything inplace` in the tests already, can you check that those do what you want?

jorisvandenbossche commented on pandas/core/arrays/boolean.py:
```diff
@@ -565,33 +568,38 @@ def logical_method(self, other):
-                warnings.filterwarnings("ignore", "elementwise", FutureWarning)
-                with np.errstate(all="ignore"):
-                    result = op(self._data, other)
+            if other_is_scalar and not isinstance(other, (type(libmissing.NA), bool)):
```
You'll need to add `np.bool_` as well I think (and tests for this), as a numpy scalar is not an instance of a bool:

```
In [14]: isinstance(np.bool_(True), bool) 
Out[14]: False
```

jorisvandenbossche replied:
Also, `other is pd.NA or isinstance(other, bool)` is a bit faster, although I am not sure we care about the 60 ns difference

jorisvandenbossche replied:
Hmm, but I see that you actually already have a test for numpy scalars?

TomAugspurger replied:
Ah, I was tripped up by `np.bool` vs `np.bool_`.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into boolean-array-kleene

TomAugspurger commented:
@jorisvandenbossche what did you do about NumPy scalars not returning NotImplemented for certain operations? We're getting the wrong answer for NumPy scalars, since our implementation isn't called.

```python
In [2]: a = pd.array([True, False, None])

In [3]: a | np.bool_(True)
Out[3]:
<BooleanArray>
[True, True, True]
Length: 3, dtype: boolean

In [4]: np.bool_(True) | a
Out[4]:
<BooleanArray>
[True, True, NA]
Length: 3, dtype: boolean
```

TomAugspurger committed:
move

TomAugspurger committed:
numpy scalars

TomAugspurger committed:
doc note

jorisvandenbossche commented:
Hmm, for the NA scalar itself I skipped the tests, as I thought there was nothing to do about this:

https://github.com/pandas-dev/pandas/blob/ee6e6b38528f88cbdc850a2c3b671825ed73581c/pandas/tests/scalar/test_na_scalar.py#L49-L69

But for the array-level ops this seems even more annoying .. (more likely to run into, as we do return numpy scalars from indexing/ops). 
But I thought that `__array_priority__` should ensure that ops are directed to the BooleanArray? (numpy scalars should have a very low array priority)

jorisvandenbossche commented:
I am actually linking to the comparison ops tests, for logical ops I apparently didn't add tests for numpy scalars (something to improve!)

TomAugspurger commented:
Ahh, we're going through `__array_ufunc__`, which is not using Kleene logic. Should have been obvious that we got control somehow, as that returned a BooleanArray.

Will fix that.

TomAugspurger committed:
handle numpy bool

jorisvandenbossche commented:
> Should have been obvious that we got control somehow, as that returned a BooleanArray.

Aha, yes missed that as well :)

We probably need to fix this for the other ops as well (can be in another PR)

TomAugspurger commented:
OK, NumPy bools should be handled now. I did that by handling `or`, `xor`, and `and` in `maybe_dispatch_ufunc_to_dunder_op`. Hopefully that doesn't break anything else.

TomAugspurger commented:
All green. Should be good to go.

jreback commented on asv_bench/benchmarks/boolean.py:
```diff
@@ -0,0 +1,32 @@
+
+import pandas as pd
+
+
```
do we have others that could be moved here? (e.g. the boolean Series ops)

TomAugspurger replied:
I don't see any in `benchmarks/binary_ops.py` or `benchmarks/series_methods.py`, though I may be missing some.

jreback commented on pandas/core/ops/mask_ops.py:
```diff
@@ -0,0 +1,178 @@
+    raise_for_nan(right, method="or")
+
+    if right is libmissing.NA:
+        result = left.copy()
```
why do you need to copy?

TomAugspurger replied:
This avoids something like

```python
>>> a = pd.array([True, False, None])
>>> b = a | pd.NA
>>> b[0] = False
```

mutating `a` inplace. We have a test for this `TestLogicalOperators.test_scalar`.

jreback reviewed (approved):
lgtm. just a couple of questions.

jreback commented:
@TomAugspurger happy to merge if @jorisvandenbossche ok

jbrockmendel commented on pandas/core/arrays/boolean.py:
```diff
@@ -184,6 +184,9 @@ class BooleanArray(ExtensionArray, ExtensionOpsMixin):
     a boolean array with the mask (True indicating missing).
 
+    BooleanArray implements Kleene logic (sometimes called three-value
+    logic) for logical operations. See :ref:`` for more.
```
is "ref" here a placeholder?

jbrockmendel commented on pandas/core/arrays/boolean.py:
```diff
@@ -565,33 +568,40 @@ def logical_method(self, other):
-                    raise ValueError("Lengths must match to compare")
                 other, mask = coerce_to_array(other, copy=False)
+            elif isinstance(other, np.bool_):
+                other = other.item()
```
this is to convert to a python bool?  why not just `bool(other)`?  `item` i usually think of as being an array method

jorisvandenbossche replied:
`item` is the general method to get a python scalar (here we of course know we want a bool).

But Tom, why is it exactly needed to convert this? I would think the numpy operations later on work fine with a numpy scalar as well?

TomAugspurger replied:
IIRC, we do things like `if right is False` or `if right is True`, which will fail for numpy booleans. I don't want to have to worry about checking both, so easier to convert here.

jbrockmendel commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -368,43 +375,211 @@ def get_op_from_name(self, op_name):
+        [
+            True,
+            False,
+            # pd.NA
```
is this a comment or commented-out code?

jbrockmendel commented:
> what did you do about NumPy scalars not returning NotImplemented for certain operations? We're getting the wrong answer for NumPy scalars, since our implementation isn't called.

Not clear that it helps _here_, but it might be relevant that pd.NA has no `__array_priority__` attr

jorisvandenbossche commented:
> it might be relevant that pd.NA has no __array_priority__ attr

It indeed doesn't help here (since the object is BooleanArray, not pd.NA, and BooleaArray already has an array_priority set). 
But, I tried it yesterday for pd.NA, as it would be nice to solve those inconsistencies. But I don't see any difference ..

So @jbrockmendel is the following correct? My understanding is that `__array_priority__` ensures that if an operation is done between two objects that both have an `__array_priority__` set, that the object with the highest priority can perform the operation first, even if it is the right hand side object? 

That can ensure that both these operations return a Timestamp:

```
In [15]: pd.Timestamp("2012-01-01") + np.timedelta64(1, 'h')
Out[15]: Timestamp('2012-01-01 01:00:00')

In [16]: np.timedelta64(1, 'h') + pd.Timestamp("2012-01-01") 
Out[16]: Timestamp('2012-01-01 01:00:00')
```

And, indeed, if I comment out `__array_priority__` on `Timestamp`, we get:

```
In [1]: pd.Timestamp("2012-01-01") + np.timedelta64(1, 'm')
Out[1]: Timestamp('2012-01-01 00:01:00')

In [2]: np.timedelta64(1, 'h') + pd.Timestamp("2012-01-01")
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
<ipython-input-2-0de73ced9bd3> in <module>
----> 1 np.timedelta64(1, 'h') + pd.Timestamp("2012-01-01")

UFuncTypeError: ufunc 'add' cannot use operands with types dtype('<m8[h]') and dtype('O')
```

But if I add `__array_priority__ = 100` to the NAType class, I still see this inconsistent behaviour:

```
In [1]: pd.NA == np.bool_(True) 
Out[1]: NA

In [2]: np.bool_(True) == pd.NA 
Out[2]: False
```

So it seems that the decision who gets priority or which operation is executed first is more complicated (or array priority is not honoured here, which is maybe something to report to numpy?)

TomAugspurger commented:
@jorisvandenbossche does it make sense to enable `__array_ufunc__`? I don't know if the NumPy scalar is invoking the ufunc machinery for binary ops but it may be. That'd also get `pd.NA` working better with raw ndarrays.

jbrockmendel commented:
> My understanding is that __array_priority__ ensures that if an operation is done between two objects that both have an __array_priority__ set, that the object with the highest priority can perform the operation first, even if it is the right hand side object?

That's right if `left` is an ndarray (maybe some other numpy types?) but not for e.g. `Series`, which has array_priority

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into boolean-array-kleene

TomAugspurger committed:
cleanup

jorisvandenbossche commented:
> That's right if left is an ndarray (maybe some other numpy types?) but not for e.g. Series, which has array_priority

@jbrockmendel it's about the case where `left` is a numpy scalar, not ndarrary or series

@TomAugspurger ah, good idea! That seems to give us more control. Will further comment on the other issue.

jorisvandenbossche commented:
@TomAugspurger Thanks a lot for this !
